src/client.rs

```diff
@@ -1,398 +1,475 @@
-                        .to_string_lossy()
-                        .into_owned(),
-                    name: CStr::from_ptr(name.as_ptr()).to_string_lossy().into_owned(),
-                    value: CStr::from_ptr(value.as_ptr())
-                        .to_string_lossy()
-                        .into_owned(),
-                    fmt_type: CStr::from_ptr(fmt_type.as_ptr())
-                        .to_string_lossy()
-                        .into_owned(),
-                }
-            })),
-            status if status > 0 => Ok(None),
-            status => Err(status.into()),
-        }
+        .expect("read_all worker thread panicked");
+        self.link.unregister(done_key);
+        settings.lock().sort_by_key(|(idx, _)| *idx);
+        errors.lock().sort_by_key(|(idx, _)| *idx);
+        (
+            settings.into_inner().into_iter().map(|e| e.1).collect(),
+            errors.into_inner().into_iter().map(|e| e.1).collect(),
+        )
     }
 
-    pub fn read_setting(
+    fn read_by_index(
         &self,
-        group: impl AsRef<str>,
-        name: impl AsRef<str>,
-    ) -> Option<Result<SettingValue, Error<ReadSettingError>>> {
-        self.read_setting_inner(group.as_ref(), name.as_ref())
+        index: u16,
+        done_rx: &Receiver<()>,
+        ctx: &Context,
+    ) -> Result<Option<Entry>, Error> {
+        let (tx, rx) = crossbeam_channel::bounded(1);
+        let key = self.link.register(move |msg: MsgSettingsReadByIndexResp| {
+            if index != msg.index {
+                return;
+            }
+            let _ = tx.try_send(Entry::try_from(msg));
+        });
+        self.sender.send(MsgSettingsReadByIndexReq {
+            sender_id: Some(SENDER_ID),
+            index,
+        })?;
+        let res = crossbeam_channel::select! {
+            recv(rx) -> msg => msg.unwrap().map(Some),
+            recv(done_rx) -> _ => Ok(None),
+            recv(ctx.timeout_rx) -> _ => Err(Error::TimedOut),
+            recv(ctx.cancel_rx) -> _ => Err(Error::Canceled),
+        };
+        self.link.unregister(key);
+        res
     }
 
     fn read_setting_inner(
-        &self,
-        group: &str,
-        name: &str,
-    ) -> Option<Result<SettingValue, Error<ReadSettingError>>> {
-        let setting = Setting::find(group, name)?;
-
-        let cgroup = match CString::new(group) {
-            Ok(cgroup) => cgroup,
-            Err(e) => return Some(Err(e.into())),
+        &mut self,
+        group: String,
+        name: String,
+        ctx: Context,
+    ) -> Result<Option<Entry>, Error> {
+        let (tx, rx) = crossbeam_channel::bounded(1);
+        let key = self.link.register({
+            let group = group.clone();
+            let name = name.clone();
+            move |msg: MsgSettingsReadResp| {
+                let fields = split_multipart(&msg.setting);
+                if fields.len() < 2 || fields[0] != group || fields[1] != name {
+                    return;
+                }
```

Comment:
Can we move this to a function it appears to be used similarly in write_setting_inner. Is this an anticipated mode? If so what settings is this behavior accounting for?

Reply:
Yeah it is, unfortunately there is no sequence/session associated with messages like fileio. So if two write requests are made at the same time, we register two `MsgSettingsWriteResp` callbacks. So you might have callback A triggered when the response intended for B comes in. So the check is to make sure that it's the response for the proper message. 

But its not 100% foolproof, like if you made two write requests for the same setting, you'd have no way to figure out which one was intended for which, at least as far as I can tell. This is why all the pub functions take `&mut self` even though we could get away with `&self`, so we can make sure only one action is performed at a time

Reply:
And r.e. the refactor, was able to factor out the check, looks a little cleaner